### PR TITLE
Feature/background vimeo video (Github #1903)

### PR DESCRIPTION
**Pull request by udidol**
## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

File: assets/dev/js/frontend/handlers/section.js [hunk 1]
```diff
@@ -71,6 +91,67 @@ class BackgroundVideo extends elementorModules.frontend.handlers.Base {
...
+
+		this.player.ready().then( () => {
+			jQuery( this.player.element ).addClass( 'elementor-background-video-embed' );
+			this.changeVideoSize();
+		} );
+	}
+
+	handleVimeoStartEndTimes( elementSettings, startTime ) {
+		// If a start time is defined, set the start time
+		if ( 0 !== startTime ) {
```
Comment (kobizz) [inline assets/dev/js/frontend/handlers/section.js]: `if ( startTime )`

File: assets/dev/js/frontend/handlers/section.js [hunk 2]
```diff
@@ -71,6 +91,67 @@ class BackgroundVideo extends elementorModules.frontend.handlers.Base {
...
+				if ( 0 === data.seconds ) {
+					this.player.setCurrentTime( startTime );
+				}
+			} );
+		}
+
+		// If start time is defined but an end time is not, go to user-defined start time at video end.
+		// Vimeo JS API has an ended event, but it never fires when infinite loop is defined, so we
+		// get the video duration (returns a promise) then use duration-0.5s as end time
+		this.player.getDuration().then( ( duration ) => {
```
Comment (kobizz) [inline assets/dev/js/frontend/handlers/section.js]: Do all this logic in one `timeupdate` listener

File: assets/dev/js/frontend/handlers/section.js [hunk 3]
```diff
@@ -117,19 +198,36 @@ class BackgroundVideo extends elementorModules.frontend.handlers.Base {
...
-		let videoLink = this.getElementSettings( 'background_video_link' );
-		const videoID = elementorFrontend.utils.youtube.getYoutubeIDFromURL( videoLink ),
-			playOnce = this.getElementSettings( 'background_play_once' );
+		let videoLink = this.getElementSettings( 'background_video_link' ),
+			videoID;
 
-		this.isYTVideo = ! ! videoID;
+		const playOnce = this.getElementSettings( 'background_play_once' );
+
+		if ( videoLink.includes( 'vimeo.com' ) ) {
```
Comment (kobizz) [inline assets/dev/js/frontend/handlers/section.js]: Check the logic

File: assets/dev/js/frontend/utils/vimeo.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
+export default class VimeoModule extends elementorModules.ViewModule {
```
Comment (kobizz) [inline assets/dev/js/frontend/utils/vimeo.js]: Create a base api importer

**Review by kobizz — CHANGES_REQUESTED** [review]

File: assets/dev/js/frontend/handlers/section.js (hunk 3 above)
Comment (udidol) [reply assets/dev/js/frontend/handlers/section.js]: It works

File: assets/dev/js/frontend/handlers/section/background-video.js [hunk 5]
```diff
@@ -22,10 +22,17 @@ export default class BackgroundVideo extends elementorModules.frontend.handlers.
 		return elements;
 	}
 
-	calcVideosSize() {
+	calcVideosSize( $video ) {
+		let aspectRatioSetting;
+
+		if ( this.isVimeoVideo ) {
+			aspectRatioSetting = $video ? $video.outerWidth() + ':' + $video.outerHeight() : '16:9';
```
Comment (kobizz) [inline assets/dev/js/frontend/handlers/section/background-video.js]: Remove this check, and instead check it in the event attachment

File: assets/dev/js/frontend/handlers/section/background-video.js [hunk 6]
```diff
@@ -117,19 +194,36 @@ export default class BackgroundVideo extends elementorModules.frontend.handlers.
...
-		const videoID = elementorFrontend.utils.youtube.getYoutubeIDFromURL( videoLink ),
-			playOnce = this.getElementSettings( 'background_play_once' );
+		let videoLink = this.getElementSettings( 'background_video_link' ),
+			videoID;
+
+		const playOnce = this.getElementSettings( 'background_play_once' );
 
-		this.isYTVideo = ! ! videoID;
+		if ( -1 !== videoLink.indexOf( 'vimeo.com' ) ) {
+			this.isVimeoVideo = true;
```
Comment (kobizz) [inline assets/dev/js/frontend/handlers/section/background-video.js]: Change it to `this.videoType`

**Review by kobizz — CHANGES_REQUESTED** [review]

File: assets/dev/js/frontend/handlers/section/background-video.js [hunk 7]
```diff
@@ -71,6 +93,64 @@ export default class BackgroundVideo extends elementorModules.frontend.handlers.
...
+				loop: ! elementSettings.background_play_once,
+				transparent: false,
+				playsinline: false,
+				background: true,
+			};
+
+		this.player = new Vimeo.Player( this.elements.$backgroundVideoContainer, vimeoOptions );
+
+		// Handle user-defined start/end times
+		this.handleVimeoStartEndTimes( elementSettings, startTime );
```
Comment (kobizz) [inline assets/dev/js/frontend/handlers/section/background-video.js]: Don't send this parameter

File: assets/dev/js/frontend/handlers/section/background-video.js [hunk 8]
```diff
@@ -71,6 +93,64 @@ export default class BackgroundVideo extends elementorModules.frontend.handlers.
...
+		// If a start time is defined, set the start time
+		if ( startTime ) {
+			this.player.on( 'play', ( data ) => {
+				if ( 0 === data.seconds ) {
+					this.player.setCurrentTime( startTime );
+				}
+			} );
+		}
+
+		// If an end time is defined, handle ending the video
```
Comment (kobizz) [inline assets/dev/js/frontend/handlers/section/background-video.js]: Move this comment to the right place

**Review by kobizz — CHANGES_REQUESTED** [review]

**Review by kobizz — APPROVED** [review]

**Pull request closed without merging**